tensorboard/plugins/profile/pod_viewer/stack_bar_chart/stack-bar-chart.ts

```diff
@@ -0,0 +1,203 @@
+    const barWidth = 50;
+    const xScaleRange = data.length * barWidth;
+    let xScale = d3.scaleBand().range([0, xScaleRange]).padding(0.4);
+    let yScale = d3.scaleLinear().range([height, 0]);
+    let colorScale = d3.scaleOrdinal<number, string>(d3.schemeCategory10)
+                       .domain([0, 19]);
+    let svg = d3.select('#' + this.id + ' #chart')
+                  .append('svg')
+                  .attr(
+                      'width',
+                      Math.max(width, xScaleRange + margin.left + margin.right))
+                  .attr('height', height + margin.top + margin.bottom)
+                  .append('g')
+                  .attr(
+                      'transform',
+                      'translate(' + margin.left + ',' + margin.top + ')');
+    let stack = d3.stack()
+                    .keys(stackKey)
+                    .order(d3.stackOrderNone)
+                    .offset(d3.stackOffsetNone);
+    const layers = stack(data);
+    xScale.domain(data.map(this.xDomainFunc));
+    yScale.domain([0, d3.max(layers[layers.length - 1], (d) => d[0] + d[1])])
+        .nice();
+    this.drawLayers(svg, layers, xScale, yScale, colorScale);
+    this.drawAxes(svg, xScale, yScale, height);
+    this.drawLegend(svg, stackLabel, colorScale);
+  },
+  /**
+   * Draw the layers for all the bars.
+   */
+  drawLayers: function(svg, layers, xScale, yScale, colorScale) {
+    let parent = this;
+    let layer = svg.selectAll('.layer')
+                    .data(layers)
+                    .enter()
+                    .append('g')
+                    .attr('class', 'layer')
+                    .style('fill', (d, i) => colorScale(i));
+    layer.selectAll('rect')
+        .data((d) => d)
+        .enter()
+        .append('rect')
+        .attr('width', xScale.bandwidth())
+        .attr('y', (d) => yScale(d[1]))
+        .attr('height', (d) => yScale(d[0]) - yScale(d[1]))
+        .attr('x', (d, i) => xScale(parent.xDomainFunc(d.data)))
+        .on('mouseover',
+            function(d) {
+              d3.select(this).style('opacity', 0.5);
+              parent.active = d.data;
+            })
+        .on('mouseout', function(d) {
+          d3.select(this).style('opacity', 1.0);
+          parent.active = null;
+        });
+  },
+  /**
+   * Draw the axes of the chart.
+   */
+  drawAxes: function(svg, xScale, yScale, height) {
+    let xAxis = d3.axisBottom(xScale);
+    let yAxis = d3.axisLeft(yScale);
+    svg.append('g')
+        .attr('class', 'axis axis--x')
+        .style('font-size', 14)
+        .attr('transform', 'translate(0,' + (height + 5) + ')')
+        .call(xAxis);
+    svg.append('g')
+        .attr('class', 'axis axis--y')
+        .style('font-size', 14)
+        .attr('transform', 'translate(0,0)')
+        .call(yAxis);
+  },
+  /**
+   * Draw the legends of the chart.
+   */
+  drawLegend: function(svg, labels, colorScale) {
+    const legendWidth = 150;
```

Comment:
these look like module level constant. Please put it on the top of the module.

Reply:
Done.